django/template/base.py

```diff
@@ -981,6 +981,16 @@ def __repr__(self):
     def render(self, context):
         return self.s
 
+    def render_annotated(self, context):
+        """
+        Special case implementation for performance reasons. The default
+        implementation handles the case where the underlying render() method
+        may throw an exception (of any type) during its execution; however for
+        text nodes specifically, not even a cast is required, so the value
+        may be immediately returned unchecked.
+        """
+        return self.s
```

Comment:
I would call `render()`, if case someone subclasses `TextNode` :thinking:  
```suggestion
        return self.render()
```

Reply:
This would be very similar to the implementation that we have today. I suspect the extra function call would make it very similar in performance terms to what we have today. 

I looked at making this change a few months ago but I couldn't evidence any performance gain when rendering something like a simple form. I found the various `if` nodes were dramatically more expensive. 

I'd be really happy to be proven wrong with a benchmark. 👍

Reply:
Yeah, I had considered that, but largely dismissed it because ... what's the point in subclassing `TextNode`? the only _reason_ to do it would be for handling how `{% extends %}` works internally ... that is, something exotic.

I haven't checked, but I _suspect_ that avoiding the method call is a substantial part of the cProfile timed difference. I could be wrong (I'll need to re-run everything), in which case sure, let's call the `render()` method.

If it turns out that avoiding the `render()` method is important, and you think it's worth the effort to deal with the potential backwards incompatibility of changing how it works (I don't, fwiw), I think I _can_ make it raise a warning or exception via `__init_subclass__`

Reply:
Possible best of both worlds?

```
class TextNode(Node):
    ...
    def __init_subclass__(cls, **kwargs):
        if cls.render_annotated is TextNode.render_annotated:
            warnings.warn('whee, deprecated blah blah, provide your own')
            cls.render_annotated = Node.render_annotated
```
and then:
```
In [1]: from django.template.base import TextNode
In [2]: class TextNode2(TextNode):  pass
UserWarning: ...
In [3]: TextNode.render_annotated
Out[3]: <function django.template.base.TextNode.render_annotated(self, context)>
In [4]: TextNode2.render_annotated
Out[4]: <function django.template.base.Node.render_annotated(self, context)>

Reply:
@kezabelle Agreed, it's not worth to raise a warning. We can keep the proposed version of `render_annotated()`.

Reply:
What do you think about looking at eeking out a bit more performance here by using `__slots__`?

Reply:
For the faster attribute access? 
It's something I'm sort of looking at generally, across the various bits of Nodes etc (eg: on `FilterExpression` and `Variable` too), amongst various other things (half of which have been dead ends :p).

I think it's probably worth doing, but I think it's probably worth tracking separately (ticket/changeset) so that any benefits observed can be documented and attributed specifically to that (though given attribute access outside of `getattr` doesn't turn up in things like cProfile, I guess walltime is maybe all we've got?)

Reply:
>  think it's probably worth doing, but I think it's probably worth tracking separately (ticket/changeset) so that any benefits observed can be documented and attributed specifically to that (though given attribute access outside of getattr doesn't turn up in things like cProfile, I guess walltime is maybe all we've got?)

see ticket-12826